src/devices/src/virtio/rng/device.rs

```diff
@@ -404,9 +483,135 @@ mod tests {
             ],
         );
 
+        let entropy_event_fails = METRICS.entropy.entropy_event_fails.count();
+        let entropy_event_count = METRICS.entropy.entropy_event_count.count();
+        let entropy_bytes = METRICS.entropy.entropy_bytes.count();
+        let host_rng_fails = METRICS.entropy.host_rng_fails.count();
         assert_eq!(th.emulate_for_msec(100).unwrap(), 1);
-        assert_eq!(METRICS.entropy.entropy_event_fails.count(), 1);
-        assert_eq!(METRICS.entropy.entropy_event_count.count(), 4);
-        assert_eq!(METRICS.entropy.entropy_bytes.count(), 542);
+        assert_eq!(
+            METRICS.entropy.entropy_event_fails.count(),
+            entropy_event_fails
+        );
+        assert_eq!(
+            METRICS.entropy.entropy_event_count.count(),
+            entropy_event_count + 1
+        );
+        assert_eq!(METRICS.entropy.entropy_bytes.count(), entropy_bytes + 512);
+        assert_eq!(METRICS.entropy.host_rng_fails.count(), host_rng_fails);
+    }
+
+    #[test]
+    fn test_bad_rate_limiter_event() {
+        let mem = create_virtio_mem();
+        let mut th = VirtioTestHelper::<Entropy>::new(&mem, default_entropy());
+
+        th.activate_device(&mem);
+        let mut dev = th.device();
+
+        check_metric_after_block!(
+            &METRICS.entropy.entropy_event_fails,
+            1,
+            dev.process_rate_limiter_event()
+        );
+    }
+
+    #[test]
+    fn test_bandwidth_rate_limiter() {
+        let mem = create_virtio_mem();
+        // Rate Limiter with 4000 bytes / sec allowance and no initial burst allowance
+        let device = Entropy::new(RateLimiter::new(4000, 0, 1000, 0, 0, 0).unwrap()).unwrap();
+        let mut th = VirtioTestHelper::<Entropy>::new(&mem, device);
+
+        th.activate_device(&mem);
+
+        // We are asking for 4000 bytes which should be available, so the
+        // buffer should be processed normally
+        th.add_desc_chain(RNG_QUEUE, 0, &[(0, 4000, VIRTQ_DESC_F_WRITE)]);
+        check_metric_after_block!(
+            METRICS.entropy.entropy_bytes,
+            4000,
+            th.device().process_entropy_queue()
+        );
+        assert!(!th.device().rate_limiter.is_blocked());
+
+        // Sleep for 1 second to completely replenish the rate limiter
+        std::thread::sleep(Duration::from_millis(1000));
```

Comment:
Nit: Could we just call `RateLimiter::manual_replenish` here? Just in case there's some weird timing conditions that could cause spurious failures.

Reply:
I dislike it too and I think in this particular case we can do it. But these things are all over the place and I'm not sure we can get rid of them. I 'm a bit worried about having the rate limiter replenishing itself and us doing `manual_replenish` will make this logic quite hard to follow.

Reply:
Ah I see, yeah I didn't take that into consideration. Makes sense to keep it this way then